### PR TITLE
Add libspirv::CapabilitySet

### PR DESCRIPTION
**Pull request by dneto0**
It's optimized for the common case, where capabilities have value
at most 63.

### CODE REVIEW

**Comment by dneto0** [comment]
There are two commits here.  The first one adds libspirv::CapabilitySet and tests it.
The second one uses CapabilitySet to store sets of capabilities (duh) instead of using a 64-bit mask.

We'll need this for the case where extensions will have capabilities with values higher than 63.

**Comment by dneto0** [comment]
Looks like I have to adapt to VisualStudio 2013 incompatibilities with std::initializer_list constructors...

**Comment by dneto0** [comment]
With lots of help from @awoloszyn we found that on VS 2013, the ValidationState member initializer for module_capabilities_({}) was causing CapabilitySet(0) to be called, and hence insert enum valued 0 into the set.  :-(
We fixed it by just using module_capabilities_()

Also, added "explicit" qualifier to the single-argument constructor for CapabilitySet.

**Comment by dneto0** [comment]
Rebased

File: source/capability_set.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+// It is fast and compact for the common case, where capability values
+// are at most 63.  But it can represent capabilities with larger values,
+// as may appear in extensions.
+class CapabilitySet {
+ private:
+  using OverflowSetType = std::set<uint32_t>;
+
+ public:
+  // Construct an empty set.
+  CapabilitySet() {}
```
Comment (antiagainst) [inline source/capability_set.h]: `= default`?

File: source/capability_set.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  // Returns true if the capability represented as a 32-bit word is in the set.
+  bool Contains(uint32_t word) const;
+
+  // Applies f to each capability in the set, in order from smallest enum
+  // value to largest.
+  void ForEach(std::function<void(SpvCapability)> f) const;
+
+ private:
+  // Returns true if the set has capabilities with value greater than 63.
+  bool HasOverflow() { return overflow_.get() != nullptr; }
```
Comment (antiagainst) [inline source/capability_set.h]: I think `.get()` is redundant here.

File: source/capability_set.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+void CapabilitySet::Add(uint32_t word) {
+  if (auto new_bits = AsMask(word)) {
+    mask_ |= new_bits;
+  } else {
+    Overflow().insert(word);
+  }
+}
+
+bool CapabilitySet::Contains(uint32_t word) const {
+  // We shouldn't call Overflow() since this is a const // method.
```
Comment (antiagainst) [inline source/capability_set.cpp]: `const method`

File: source/capability_set.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  return false;
+}
+
+// Applies f to each capability in the set, in order from smallest enum
+// value to largest.
+void CapabilitySet::ForEach(std::function<void(SpvCapability)> f) const {
+  for (uint32_t i = 0; i < 64; ++i) {
+    if (mask_ & AsMask(i)) f(static_cast<SpvCapability>(i));
+  }
+  if (overflow_.get()) {
```
Comment (antiagainst) [inline source/capability_set.cpp]: I think `.get()` is redundant here.

File: source/capability_set.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+namespace libspirv {
+
+// A set of values of type SpvCapability.
+// It is fast and compact for the common case, where capability values
+// are at most 63.  But it can represent capabilities with larger values,
+// as may appear in extensions.
+class CapabilitySet {
+ private:
+  using OverflowSetType = std::set<uint32_t>;
```
Comment (antiagainst) [inline source/capability_set.h]: `unordered_set`?

File: utils/generate_grammar_tables.py [hunk 6]
```diff
@@ -59,10 +37,7 @@ def compose_capability_mask(caps):
     Returns:
       a string containing the hexadecimal value of the bit mask
     """
-    assert len(CAPABILITY_BIT_MAPPING) != 0
-    bits = [CAPABILITY_BIT_MAPPING[c] for c in caps]
-    caps_mask = functools.reduce(lambda m, b: m | (1 << b), bits, 0)
-    return '0x{:04x}'.format(caps_mask)
+    return "{" + ", ".join(['SpvCapability{}'.format(c) for c in caps]) + "}"
```
Comment (antiagainst) [inline utils/generate_grammar_tables.py]: I'd suggest to use `"...".format()`. This is kinda hard to read.

File: utils/generate_grammar_tables.py [hunk 7]
```diff
@@ -31,26 +28,7 @@ def make_path_to_file(f):
...
-            print(
-                "error: capability enumerant {} valued {} is over 63; "
-                "spv_capability_mask_t doesn't support this".format(
-                    enumerant['enumerant'], enumerant['value']))
-            exit(1)
-        CAPABILITY_BIT_MAPPING[enumerant['enumerant']] = enumerant['value']
-
-
-def compose_capability_mask(caps):
+def compose_capability_list(caps):
```
Comment (antiagainst) [inline utils/generate_grammar_tables.py]: Nice to see complicated logic got killed! ;)

File: utils/generate_grammar_tables.py [hunk 8]
```diff
@@ -59,10 +37,7 @@ def compose_capability_mask(caps):
     Returns:
       a string containing the hexadecimal value of the bit mask
```
Comment (antiagainst) [inline utils/generate_grammar_tables.py]: This needs to be updated.

File: utils/generate_grammar_tables.py (hunk 6 above)
Comment (antiagainst) [inline utils/generate_grammar_tables.py]: Well, never mind. Just my eyes. ;)

File: source/capability_set.h (hunk 5 above)
Comment (dneto0) [inline source/capability_set.h]: The ForEach method promises to call the capabilities in order.  So I want an ordered set here. I'll add the comment saying so.

**Comment by dneto0** [comment]
Rebased and addressed all comments.

**Comment by antiagainst** [comment]
LGTM

**Comment by dneto0** [comment]
Rebased

**Pull request merged**